### PR TITLE
SI-7350 Prevent resource leaks in PipedProcesses.runAndExitValue()

### PR DESCRIPTION
**Pull request by rubyu**
This patch fixes some problems in `PipedProcesses` as listed bellow:
- `source` and `sink` leak when b.run() fails at [ProcessImpl.scala#L132](https://github.com/scala/scala/blob/538aa22ec8144430e778a34bc749eccb6f5dbf4c/src/library/scala/sys/process/ProcessImpl.scala#L132). ([SI-7350](https://issues.scala-lang.org/browse/SI-7350))
- `source`, `sink` and `second` leak when a.run() fails at [ProcessImpl.scala#L133](https://github.com/scala/scala/blob/538aa22ec8144430e778a34bc749eccb6f5dbf4c/src/library/scala/sys/process/ProcessImpl.scala#L133). ([SI-7350](https://issues.scala-lang.org/browse/SI-7350))
- `source` and `sink` leak when `Process.exitValue()` interrupted in [ProcessImpl.scala#L135](https://github.com/scala/scala/blob/538aa22ec8144430e778a34bc749eccb6f5dbf4c/src/library/scala/sys/process/ProcessImpl.scala#L135). ([SI-7350](https://issues.scala-lang.org/browse/SI-7350))
- `first` do not properly connected with `second` when `None` putted to `currentSink` before `sink` receive `Some(InputStream)` at [ProcessImpl.scala#L138](https://github.com/scala/scala/blob/538aa22ec8144430e778a34bc749eccb6f5dbf4c/src/library/scala/sys/process/ProcessImpl.scala#L138). ([SI-8768](https://issues.scala-lang.org/browse/SI-8768))

and, a broblem in [BasicIO.scala#L222](https://github.com/scala/scala/blob/538aa22ec8144430e778a34bc749eccb6f5dbf4c/src/library/scala/sys/process/BasicIO.scala#L222) that 
- `InputStream` and `OutputStream` never throw `InterruptedException`, but `BasicIO.transferFully` trying to catch it at [package.scala#L238](https://github.com/scala/scala/blob/159baa268e5cc14b50bf29a40fff922e36dd7bc9/src/library/scala/sys/process/package.scala#L238). 

I guess this originally intented to catch `InterruptedIOException`.

### CODE REVIEW

**Comment by Ichoran** [comment]
LGTM though I can't claim to have followed every piece of logic to the very end.  Nice tests, by the way!

File: test/files/run/t7350.scala [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+    }
+    Await.result(f, Duration(5, SECONDS))
+    println(s"- in.closed: ${in.closed}")
+    println(s"- out.closed: ${out.closed}")
+    println(s"- source.isReleased: ${source.isReleased}")
+    println(s"- sink.isReleased: ${sink.isReleased}")
+  }
+}
+
+package scala.sys.process {
```
Comment (gkossakowski) [inline test/files/run/t7350.scala]: Why does debug object need to live in `process` package?

**Comment by gkossakowski** [comment]
Hi @rubyu! Thanks for excellent patch!

Have you considered writing your tests as junit tests?

**Comment by rubyu** [comment]
I only imitated https://github.com/scala/scala/pull/2379. Should I rewrite this patch to use junit?

File: test/files/run/t7350.scala (hunk 1 above)
Comment (rubyu) [inline test/files/run/t7350.scala]: This debug object is just only for not to conflict class declarations and easy to import test target classes. Is this bad manners?

Comment (gkossakowski) [inline test/files/run/t7350.scala]: I'd prefer to not inject code into existing packages unless it's absolutely necessary.

**Comment by gkossakowski** [comment]
There are two issues:
1. Unit tests should be written using junit if possible
2. The [comment](https://github.com/scala/scala/pull/2379#issuecomment-18090797) from #2379 still stands: time-based tests are bad idea due to large variance in testing environment.

What's the rationale for using futures in first place?

File: test/files/run/t7350.scala (hunk 1 above)
Comment (rubyu) [inline test/files/run/t7350.scala]: The classes are `private[process]`, then I create customized and mocked classes of it in `process.debug`, I could not find any other way...

**Comment by rubyu** [comment]
I'm using `Await.result` for minimizing the risks of wasting time; for example, in the future, if some changes are given to `SinkSource` and `source.release()` to be no response, then the test verifying it will never end. `Await.result` assures each test ends within the time specified. But practically, all of the tests ends in a moment. Are these considerations unwanted?

**Comment by som-snytt** [comment]
@som-snytt might want to look and admire the test later today. +1 to junit.

**Comment by rubyu** [comment]
@gkossakowski I translated the test script into junit test cases. And, should I not use `Future` and `Await.result` in the test?

**Comment by gkossakowski** [comment]
LGTM

I agree it's unlikely we'll ever have problems with using `Await.result` because those tests are very quick to run. Thanks for JUnit translation!

**Comment by gkossakowski** [comment]
@rubyu: could you drop a comment on [SI-7350](https://issues.scala-lang.org/browse/SI-7350) so you can take the credit for fixing the issue?

**Comment by rubyu** [comment]
@gkossakowski thanks, I wrote a comment.

**Comment by rubyu** [comment]
@gkossakowski I think SI-8768 would also be fixed by this pull request.

**Comment by gkossakowski** [comment]
Thanks for mentioning the other ticket. It's closed already. I'll fix assignee a bit later. For some reason, JIRA doesn't let me assign those tickets to you.

**Comment by gourlaysama** [comment]
@gkossakowski you have to be member of the `jira-developers` group to be assigned an issue, but people are not in by default.

**Comment by gkossakowski** [comment]
Thanks @gourlaysama! All fixed now :+1:

**Pull request merged**